### PR TITLE
Otel integration test apps

### PR DESCRIPTION
**Pull request by luismiramirez**
This PR contains a proposal of an integration test architecture for all the supported instrumentations. If approved, it'll be our standard and will open a new PR for each instrumentation combination we already have. After that, we'll add Postgres and GraphQL instrumentations.

---
## [Remove old integration test helpers](https://github.com/appsignal/appsignal-nodejs/commit/5fc3b4020c8ad8fb354215bf1824c1bf4eede618)

Old `test/integration/support` are no longer needed for the
OpenTelemetry version of the integration.

## [Add testMode and testModeFilePath config opts](https://github.com/appsignal/appsignal-nodejs/commit/298a4a164167f0bb2ad6d2d62d7fc22f9f07d0b0)

When in test mode (`testMode:true`) the `TestModeSpanProcessor` is added
to the tracer provider and it saves spans in a JSON file (path based on
the contents of `testModeFilePath`) for the integration tests to read.

## [Add Express + Redis test app](https://github.com/appsignal/appsignal-nodejs/commit/d5bd6ba37f1553e7de48183cb226a21d598438bf)

This commit adds a test app using Express and Redis plus the structure
future test apps will have.

It includes a module with common methods that will be shared in all the
apps.

To run the tests position in `test/express-redis/` and run
`docker-compose up`.

## [Add Express Redis integration tests to CI](https://github.com/appsignal/appsignal-nodejs/commit/24fe23404a3e45343c145c9f11aa8500f6aa375d)

This commit adds a script to run integration test apps and the addition
of the Express Redis test app to the Semaphore build matrix.

## [Comply with linter suggestions](https://github.com/appsignal/appsignal-nodejs/commit/e486e2ff3cf291d4e6568eecf3c780973f3cc24e)

Due to the nature of the PR, some minor linter offenses needed to be
addressed.

## [Stop integration tests on container exit](https://github.com/appsignal/appsignal-nodejs/commit/50c5bbca61a409c8c20728c7b024ed44e14ec44a)

Integration tests are run in their own container so docker-compose needs
to know when to exit to tell the CI the result code.

This commit adds the flags --abort-on-container-exit to exit with code >
0 when something goes wrong and --exit-code-from tests to base the

[skip changeset]

### CODE REVIEW

File: test/integration_helper.rb
Comment (luismiramirez) [inline test/integration_helper.rb:62]: I'll keep all these instrumentation-specific helpers here until we have all the apps in place, because I can see all of them being reused. Even the Redis ones.

File: src/config/options.ts
Comment (tombruijn) [inline src/config/options.ts:34]: Other test specific config we've only added as env vars, like `APPSIGNAL_DIAGNOSE_ENDPOINT` (should have a `_` prefix) for pointing to the local endpoint in the test suite. I don't think this needs to be part of the config class because it doesn't need to be configurable by our users. You can use a `_APPSIGNAL_TEST_MODE` and `_APPSIGNAL_TEST_MODE_FILE_PATH` env var for this.

File: src/client.ts
Comment (tombruijn) [inline src/client.ts:208]: I was a bit unsure what was tested here, because this doesn't use our real SpanProcessor. It's testing all the automatically added instrumentation up to the SpanProcessor part, right?

Like I mentioned in an earlier call, we can also call `toObject()` on an OpenTelemetry Span in the SpanProcessor when this test mode is active and write that extension parsed result to a file.

File: test/express-redis/tests/Dockerfile
Comment (unflxw) [reply test/express-redis/tests/Dockerfile:10]: @tombruijn There is a point to copying the Gemfile and its lock first, then the rest of the application. It allows the Docker cache to reuse the `RUN bundle install` layer if the Gemfile and its lock haven't changed.

File: src/config/options.ts
Comment (unflxw) [reply src/config/options.ts:34]: If it's not declared on the config type, then TS won't let you access it from `this.config.data`. I don't think our config object currently supports having configuration options that can't be set via the constructor.

Regarding the `_` prefix at the beginning of the environment variable, I thought that was used only for variables that are passed on to the agent.

We could, of course, just read `process.env["APPSIGNAL_TEST_MODE"]` where needed, instead of getting it from the config object, if that's preferable.

File: test/express-redis/app/run.sh
Comment (luismiramirez) [reply test/express-redis/app/run.sh:23]: I think it's best to remove the `file://..` and keep the link.

File: test/express-redis/tests/Dockerfile
Comment (tombruijn) [reply test/express-redis/tests/Dockerfile:10]: Ok, but that's only locally. On Semaphore it doesn't cache Docker layers.

If we use a volume we could cache the bundle directory on the CI and locally, so it doesn't need to download all gems all the time. We'd take small performance hit because it's mounted on a volume, but not as bad as on local macOS machines.

File: src/config/options.ts
Comment (tombruijn) [reply src/config/options.ts:34]: > Regarding the _ prefix at the beginning of the environment variable, I thought that was used only for variables that are passed on to the agent.

It's for private env vars, which include the extension's env vars.

> We could, of course, just read process.env["APPSIGNAL_TEST_MODE"] where needed, instead of getting it from the config object, if that's preferable.

I would prefer it, just because it's not a config option any of our users need to be able to set. (That they can potentially via the env var is a minimal risk.)

File: src/client.ts
Comment (luismiramirez) [reply src/client.ts:210]: I'll take some time to try the `toObject()` approach.

Comment (tombruijn) [reply src/client.ts:210]: No need to block the PR on it. You can look at it in another PR if you want.

File: src/instrumentation/express/error_handler.ts
Comment (tombruijn) [inline src/instrumentation/express/error_handler.ts:1]: This branch branched of the OpenTelemetry branch before #715 was merged so this commit shouldn't be here. Can you rebase on the latest OpenTelemetry branch state?

**Review by tombruijn — APPROVED** [review]
other than that one commit that shouldn't be in this PR, looks good

File: package.json
Comment (tombruijn) [inline package.json:21]: wait why is npm itself a dependency here? I see this directly after hitting approve.

Can you remove it?

File: src/client.ts
Comment (luismiramirez) [reply src/client.ts:210]: As talked earlier this week, I think we should keep this out of scope as the plans are relying on the upcoming HTTP server in the agent and get the spans from there.

**Review by unflxw — APPROVED** [review]

**Pull request merged**
